ooni/agent/scheduler.py

```diff
@@ -0,0 +1,409 @@
+            measurement_path.child(measurement['id']).remove()
+            amount_deleted += measurement['size']
+
+
+class RunDeck(ScheduledTask):
+    """
+    This will run the decks that have been configured on the system as the
+    decks to run by default.
+    """
+
+    def __init__(self, director, deck_id, schedule):
+        self.deck_id = deck_id
+        self.director = director
+        # We use as identifier also the schedule time
+        identifier = 'run-deck-' + deck_id + '-' + md5(schedule).hexdigest()
+        super(RunDeck, self).__init__(schedule, identifier)
+
+    @defer.inlineCallbacks
+    def task(self):
+        deck = deck_store.get(self.deck_id)
+        yield deck.setup()
+        yield deck.run(self.director)
+
+
+class RefreshDeckList(ScheduledTask):
+    """
+    This task is configured to refresh the list of decks that are enabled.
+    """
+    identifier = 'refresh-deck-list'
+    schedule = '@hourly'
+
+    def __init__(self, scheduler, schedule=None, identifier=None):
+        self.scheduler = scheduler
+        super(RefreshDeckList, self).__init__(schedule, identifier)
+
+    def first_run(self):
+        """
+        On first run we enable the default decks.
+        """
+        for deck_id in DEFAULT_DECKS:
+            deck_store.enable(deck_id)
+
+    def task(self):
+        self.scheduler.refresh_deck_list()
+
+
+class SendHeartBeat(ScheduledTask):
+    """
+    This task is used to send a heartbeat that the probe is still alive and
+    well.
+    """
+    identifier = 'send-heartbeat'
+    schedule = '@hourly'
+
+    def task(self):
+        # XXX implement this
+        pass
+
+# Order mattters
+SYSTEM_TASKS = [
+    UpdateInputsAndResources
+]
+
+
+@defer.inlineCallbacks
+def run_system_tasks(no_input_store=False):
+    task_classes = SYSTEM_TASKS[:]
+
+    if no_input_store:
+        log.debug("Not updating the inputs")
+        task_classes.remove(UpdateInputsAndResources)
+
+    for task_class in task_classes:
+        task = task_class()
+        log.debug("Running task {0}".format(task.identifier))
+        try:
+            yield task.run()
+        except DidNotRun:
+            log.debug("Did not run {0}".format(task.identifier))
+        except Exception as exc:
+            log.err("Failed to run task {0}".format(task.identifier))
+            log.exception(exc)
+
+
+class SchedulerService(service.MultiService):
+    """
+    This service is responsible for running the periodic tasks.
+    """
+    def __init__(self, director, interval=30):
+        service.MultiService.__init__(self)
+        self.director = director
+        self.interval = interval
+        self._looping_call = LoopingCall(self._should_run)
+        self._scheduled_tasks = []
+
+    def schedule(self, task):
+        self._scheduled_tasks.append(task)
+
+    def unschedule(self, task):
+        # We first cancel the task so the run lock is deleted
+        task.cancel()
```

Comment:
If the task is running it will actually continue running until the end. There is currently no concept of actually stopping a running task and things get even more tricky when a deck is concerned.

There is a ticket open regarding this, but it's deferred to future versions: https://github.com/TheTorProject/ooni-probe/issues/590